### PR TITLE
[TIPC] Adjust the order of set_device

### PR DESCRIPTION
**Pull request by YanhuiDua**
1. Adjust the order of set_device. The set_device function should be placed before build_dataset;
2. Add `"npu" in paddle.device.get_all_custom_device_type():` in main.py to support custom device.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -24,36 +24,35 @@
...
+                        '--weights',
+                        type=str,
+                        help='weights for finetuning or testing')
+    parser.add_argument('--fleet',
+                        action='store_true',
+                        help='whether to use fleet run distributed training')
+    parser.add_argument('--amp',
+                        action='store_true',
+                        help='whether to open amp training.')
     parser.add_argument(
```
Comment (YanhuiDua) [inline main.py:56]: format modification

File: main.py [hunk 2]
```diff
@@ -90,7 +89,8 @@ def main():
     cfg = get_config(args.config, overrides=args.override)
 
     # enable to use npu if paddle is built with npu
-    if paddle.device.is_compiled_with_npu():
+    if paddle.device.is_compiled_with_npu(
+    ) or "npu" in paddle.device.get_all_custom_device_type():
```
Comment (YanhuiDua) [inline main.py:93]: support npu device of custom_device

File: main.py [hunk 3]
```diff
@@ -123,19 +123,19 @@ def main():
     elif args.train_dali:
         train_dali(cfg, weights=args.weights, parallel=parallel)
     elif args.multigrid:
-        train_model_multigrid(
-            cfg, world_size=world_size, validate=args.validate)
+        train_model_multigrid(cfg,
+                              world_size=world_size,
+                              validate=args.validate)
     else:
```
Comment (YanhuiDua) [inline main.py:129]: format modification

File: main.py [hunk 4]
```diff
@@ -123,19 +123,19 @@ def main():
...
-            profiler_options=args.profiler_options)
+        train_model(cfg,
+                    weights=args.weights,
+                    parallel=parallel,
+                    validate=args.validate,
+                    use_fleet=args.fleet,
+                    use_amp=args.amp,
+                    amp_level=args.amp_level,
+                    max_iters=args.max_iters,
+                    profiler_options=args.profiler_options)
```
Comment (YanhuiDua) [inline main.py:139]: format modification

File: paddlevideo/tasks/test.py [hunk 5]
```diff
@@ -32,6 +32,14 @@ def test_model(cfg, weights, parallel=True):
...
 
     """
+
+    if cfg.get('use_npu', False):
+        paddle.set_device('npu')
+    elif cfg.get('use_xpu', False):
+        places = paddle.set_device('xpu')
+    else:
+        places = paddle.set_device('gpu')
+
```
Comment (YanhuiDua) [inline paddlevideo/tasks/test.py:42]: put set_device before build_dataset. build_dataset function will call sync_stream.

File: paddlevideo/tasks/test.py [hunk 6]
```diff
@@ -45,26 +53,18 @@ def test_model(cfg, weights, parallel=True):
...
-        batch_size=batch_size,
-        num_workers=num_workers,
-        places=places,
-        drop_last=False,
-        shuffle=False)
+    dataloader_setting = dict(batch_size=batch_size,
+                              num_workers=num_workers,
+                              places=places,
+                              drop_last=False,
+                              shuffle=False)
```
Comment (YanhuiDua) [inline paddlevideo/tasks/test.py:63]: format modification

File: paddlevideo/tasks/test.py [hunk 7]
```diff
@@ -45,26 +53,18 @@ def test_model(cfg, weights, parallel=True):
...
 
     data_loader = build_dataloader(
-        dataset,
-        **dataloader_setting) if cfg.model_name not in ['CFBI'] else dataset
+        dataset, **dataloader_setting) if cfg.model_name not in ['CFBI'
+                                                                 ] else dataset
```
Comment (YanhuiDua) [inline paddlevideo/tasks/test.py:67]: format modification

**Review by qili93 — APPROVED** [review]
LGTM

**Comment by huangjun12** [comment]
LGTM

**Pull request merged**